From 3ef6aaf512c881b98b156da7fd91357483241e3d Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Thu, 21 May 2020 14:57:23 +0800
Subject: [PATCH] PONRTSYS-7437:XPCS: serdes RX EQ configuration

Serdes RX EQ settings are configurable related
to transmission line on the PCD board design,
Switching beetween speeds requires new serdes
configuration.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 drivers/net/ethernet/lantiq/xpcs/xpcs.c | 525 ++++++++++++++++++++++++--------
 drivers/net/ethernet/lantiq/xpcs/xpcs.h | 135 ++++++--
 2 files changed, 497 insertions(+), 163 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
index 9e120d79181f..8c8064fd57f3 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
@@ -33,7 +33,6 @@
 #define XPCS_RESET_NAME			"xpcs_reset"
 #define POWER_SAVE_MODE			"power-save"
 #define XPCS_MAC_IDX			"mac_idx"
-#define XPCS_SERDES_CFG			"serdes_cfg"
 #define XPCS_NO_PHY_RX_AUTO_ADAPT	"xpcs-no-phy-rx-auto-adapt"
 #define XPCS_AFE_EN			"xpcs-afe_en"
 #define XPCS_DFE_EN			"xpcs-dfe_en"
@@ -43,12 +42,15 @@
 #define XPCS_CTLE_POLE			"xpcs-rx_ctle_pole"
 #define XPCS_VGA2_GAIN			"xpcs-rx_vga2_gain"
 #define XPCS_VGA1_GAIN			"xpcs-rx_vga1_gain"
+#define XPCS_TX_EQ_PRE		"xpcs-tx_eq_pre"
+#define XPCS_TX_EQ_POST		"xpcs-tx_eq_post"
+#define XPCS_TX_EQ_MAIN		"xpcs-tx_eq_main"
+#define XPCS_TX_IBOOST_LVL		"xpcs-tx_iboost_lvl"
+#define XPCS_TX_VBOOST_LVL		"xpcs-tx_vboost_lvl"
+#define XPCS_TX_VBOOST_EN		"xpcs-tx_vboost_en"
 #define LINK_POLL_TIME			"link_poll_interval"
 #define LINK_MODE_CFG			"xpcs_mode_cfg_seq"
 
-#define PHY_BOOST(BOOST, POLE, VGA2, VGA1) \
-		((BOOST << 0) | (POLE << 5) | (VGA2 << 8) | (VGA1 << 12))
-
 static void xpcs_cl37_an(struct xpcs_prv_data *pdata);
 static void xpcs_cl73_an(struct xpcs_prv_data *pdata);
 static void xpcs_cl72_startup(struct xpcs_prv_data *pdata);
@@ -60,6 +62,104 @@ static int xpcs_2p5g_sgmii_mode(struct xpcs_prv_data *pdata);
 static void xpcs_link_poll_work(struct work_struct *work);
 static int xpcs_synphy_reset_sts(struct xpcs_prv_data *pdata);
 
+static const u32 xpcs_serdes_afe_en[] = {
+	XPCS_TENG_KR_AFE_EN,
+	XPCS_TENG_XAUI_AFE_EN,
+	XPCS_ONEG_AFE_EN,
+	XPCS_TWOP5G_AFE_EN,
+};
+
+static const u32 xpcs_serdes_dfe_en[] = {
+	XPCS_TENG_KR_DFE_EN,
+	XPCS_TENG_XAUI_DFE_EN,
+	XPCS_ONEG_DFE_EN,
+	XPCS_TWOP5G_DFE_EN,
+};
+
+static const u32 xpcs_serdes_cont_adapt_0[] = {
+	XPCS_TENG_KR_CONT_ADAPT_0,
+	XPCS_TENG_XAUI_CONT_ADAPT_0,
+	XPCS_ONEG_CONT_ADAPT_0,
+	XPCS_TWOP5G_CONT_ADAPT_0,
+};
+
+static const u32 xpcs_serdes_rx_attn_ctrl[] = {
+	XPCS_TENG_KR_PMA_RX_ATTN_CTRL,
+	XPCS_TENG_XAUI_PMA_RX_ATTN_CTRL,
+	XPCS_ONEG_PMA_RX_ATTN_CTRL,
+	XPCS_TWOP5G_PMA_RX_ATTN_CTRL,
+};
+
+static const u32 xpcs_serdes_ctrl_boost_0[] = {
+	XPCS_TENG_KR_PMA_RX_CTLE_BOOST_0,
+	XPCS_TENG_XAUI_PMA_RX_CTLE_BOOST_0,
+	XPCS_ONEG_PMA_RX_CTLE_BOOST_0,
+	XPCS_TWOP5G_PMA_RX_CTLE_BOOST_0,
+};
+
+static const u32 xpcs_serdes_ctrl_pole_0[] = {
+	XPCS_TENG_KR_PMA_RX_EQ_CTRL0,
+	XPCS_TENG_XAUI_PMA_RX_EQ_CTRL0,
+	XPCS_TWOP5G_PMA_RX_EQ_CTRL0,
+	XPCS_ONEG_PMA_RX_EQ_CTRL0,
+};
+
+static const u32 xpcs_serdes_vga2_gain_0[] = {
+	XPCS_TENG_KR_PMA_RX_VGA2_GAIN_0,
+	XPCS_TENG_XAUI_PMA_RX_VGA2_GAIN_0,
+	XPCS_TWOP5G_PMA_RX_VGA2_GAIN_0,
+	XPCS_ONEG_PMA_RX_VGA2_GAIN_0,
+};
+
+static const u32 xpcs_serdes_vga1_gain_0[] = {
+	XPCS_TENG_KR_PMA_RX_VGA1_GAIN_0,
+	XPCS_TENG_XAUI_PMA_RX_VGA1_GAIN_0,
+	XPCS_TWOP5G_PMA_RX_VGA1_GAIN_0,
+	XPCS_ONEG_PMA_RX_VGA1_GAIN_0,
+};
+
+static const u32 xpcs_serdes_tx_eq_main[] = {
+	XPCS_TENG_KR_PMA_TX_EQ_MAIN,
+	XPCS_TENG_XAUI_PMA_TX_EQ_MAIN,
+	XPCS_TWOP5G_PMA_TX_EQ_MAIN,
+	XPCS_ONEG_PMA_TX_EQ_MAIN,
+};
+
+static const u32 xpcs_serdes_tx_eq_pre[] = {
+	XPCS_TENG_KR_PMA_TX_EQ_PRE,
+	XPCS_TENG_XAUI_PMA_TX_EQ_PRE,
+	XPCS_TWOP5G_PMA_TX_EQ_PRE,
+	XPCS_ONEG_PMA_TX_EQ_PRE,
+};
+
+static const u32 xpcs_serdes_tx_eq_post[] = {
+	XPCS_TENG_KR_PMA_TX_EQ_POST,
+	XPCS_TENG_XAUI_PMA_TX_EQ_POST,
+	XPCS_TWOP5G_PMA_TX_EQ_POST,
+	XPCS_ONEG_PMA_TX_EQ_POST,
+};
+
+static const u32 xpcs_serdes_tx_iboost_lvl[] = {
+	XPCS_TENG_KR_PMA_TX_EQ_IBOOST_LVL,
+	XPCS_TENG_XAUI_PMA_TX_EQ_IBOOST_LVL,
+	XPCS_TWOP5G_PMA_TX_EQ_IBOOST_LVL,
+	XPCS_ONEG_PMA_TX_EQ_IBOOST_LVL,
+};
+
+static const u32 xpcs_serdes_tx_vboost_lvl[] = {
+	XPCS_TENG_KR_PMA_TX_EQ_VBOOST_LVL,
+	XPCS_TENG_XAUI_PMA_TX_EQ_VBOOST_LVL,
+	XPCS_TWOP5G_PMA_TX_EQ_VBOOST_LVL,
+	XPCS_ONEG_PMA_TX_EQ_VBOOST_LVL,
+};
+
+static const u32 xpcs_serdes_vboost_en[] = {
+	XPCS_TENG_KR_PMA_TX_EQ_VBOOST_EN,
+	XPCS_TENG_XAUI_PMA_TX_EQ_VBOOST_EN,
+	XPCS_TWOP5G_PMA_TX_EQ_VBOOST_EN,
+	XPCS_ONEG_PMA_TX_EQ_VBOOST_EN,
+};
+
 struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 	{
 		.id = TENG_KR_MODE,
@@ -70,9 +170,6 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mplla_bw = 0x7B,
 		.vco_ld_val[0] = 0x549,
 		.vco_ref_ld[0] = 0x29,
-		.afe_en = 1,
-		.dfe_en = 1,
-		.cont_adapt0 = 1,
 		.tx_rate[0] = 0,
 		.rx_rate[0] = 0,
 		.tx_width[0] = 2,	/* 2'b10 - 16 bit */
@@ -80,15 +177,10 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mplla_div165_clk_en = 1,
 		.mplla_div10_clk_en = 0,
 		.mplla_div8_clk_en = 1,
-		.tx_eq_main = 0x1E,
-		.tx_eq_pre = 0x14,
-		.tx_eq_post = 0x14,
 		.tx_eq_ovrride = 1,
 		.ref_clk_ctrl = 0x171,
 		.los_thr = 4,
-		.phy_boost_gain_val = 0x774a,
 		.rx_vref_ctrl = 0,
-		.vboost_lvl = 0x5,
 		.tx_iboost[0] = 0xF,
 		.set_mode = xpcs_teng_kr_mode,
 	},
@@ -103,11 +195,8 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mpllb_bw = 0x56,
 		.vco_ld_val[0] = 0x0550,
 		.vco_ref_ld[0] = 0x22,
-		.afe_en = 0,
 		.afe_en_31 = 0,
-		.dfe_en = 0,
 		.dfe_en_31 = 0,
-		.cont_adapt0 = 0,
 		.tx_rate[0] = 2,	/* 2'b10 */
 		.rx_rate[0] = 2,	/* 2'b10 */
 		.tx_width[0] = 1,
@@ -118,15 +207,10 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mpllb_div_clk_en = 0,
 		.mpllb_div10_clk_en = 1,
 		.mpllb_div8_clk_en = 0,
-		.tx_eq_main = 0x28,
-		.tx_eq_pre = 0,
-		.tx_eq_post = 0,
 		.tx_eq_ovrride = 1,
 		.ref_clk_ctrl = 0x171,
 		.los_thr = 1,
-		.phy_boost_gain_val = 0x7706,
 		.rx_vref_ctrl = 0xF,	/* Setting BIT12:8 = 5'd15 */
-		.vboost_lvl = 0x5,
 		.tx_iboost[0] = 0xF,
 		.set_mode = xpcs_teng_xaui_mode,
 	},
@@ -139,9 +223,6 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mplla_bw = 0x6E,
 		.vco_ld_val[0] = 0x0540,
 		.vco_ref_ld[0] = 0x002a,
-		.afe_en = 0,
-		.dfe_en = 0,
-		.cont_adapt0 = 0,
 		.tx_rate[0] = 3,
 		.rx_rate[0] = 3,
 		.tx_width[0] = 1,	/* 2'b01 - 10 bit */
@@ -149,16 +230,11 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mplla_div165_clk_en = 0,
 		.mplla_div10_clk_en = 1,
 		.mplla_div8_clk_en = 0,
-		.tx_eq_main = 0x28,
-		.tx_eq_pre = 0,
-		.tx_eq_post = 0,
 		.tx_eq_ovrride = 1,
 		.los_thr = 0,
 		.ref_clk_ctrl = 0x1F1,
 		.los_thr = 4,
-		.phy_boost_gain_val = 0x7706,
 		.rx_vref_ctrl = 0xF,	/* Setting BIT12:8 = 5'd15 */
-		.vboost_lvl = 0x5,
 		.tx_iboost[0] = 0xF,
 		.set_mode = xpcs_oneg_xaui_mode,
 	},
@@ -174,11 +250,8 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.vco_ld_val[0] = 0x550,
 		.vco_ref_ld[0] = 0x22,
 		.vco_ld_val[1] = 0x550,
-		.afe_en = 0,
-		.dfe_en = 0,
 		.afe_en_31 = 0,
 		.dfe_en_31 = 0,
-		.cont_adapt0 = 0,
 		.cont_adapt31 = 0,
 		.tx_rate[0] = 2, /* 3'b010 : baud/4 */
 		.rx_rate[0] = 2, /* 2'b10 : baud/4 */
@@ -192,16 +265,11 @@ struct xpcs_mode_cfg mode_cfg[MAX_XPCS_MODE] = {
 		.mpllb_div10_clk_en = 1,
 		.mpllb_div8_clk_en = 0,
 		.rx_div165_clk_en = 0,
-		.tx_eq_main = 0x13,
-		.tx_eq_pre = 4,
-		.tx_eq_post = 4,
 		.tx_eq_ovrride = 1,
 		.los_thr = 1,
 		.ref_clk_ctrl = 0x1F1,
 		.los_thr = 1,
-		.phy_boost_gain_val = 0x7706,
 		.rx_vref_ctrl = 0xF,	/* Setting BIT12:8 = 5'd15 */
-		.vboost_lvl = 0x5,
 		.tx_iboost[0] = 0xF,
 		.set_mode = xpcs_2p5g_sgmii_mode,
 	},
@@ -392,6 +460,7 @@ static int xpcs_synphy_reset_sts(struct xpcs_prv_data *pdata)
 
 static void xpcs_cfg_table(struct xpcs_prv_data *pdata)
 {
+	struct xpcs_serdes_cfg *serdes_cfg = pdata->serdes_cfg;
 	u32 mplla_ctrl2 = 0;
 	u32 tx_eq0, tx_eq1;
 	u32 val = 0, i = 0;
@@ -479,7 +548,7 @@ static void xpcs_cfg_table(struct xpcs_prv_data *pdata)
 		XPCS_RGWR_VAL(pdata, PMA_RX_GENCTRL2, RX_WIDTH_0,
 			      pdata->mode_cfg->rx_width[0]);
 		XPCS_RGWR_VAL(pdata, PMA_TX_BOOST_CTRL, TX0_IBOOST,
-			      pdata->mode_cfg->tx_iboost[0]);
+			      serdes_cfg->tx_iboost_lvl[pdata->mode]);
 
 		break;
 
@@ -490,18 +559,18 @@ static void xpcs_cfg_table(struct xpcs_prv_data *pdata)
 	val = XPCS_RGRD(pdata, PMA_AFE_DFE_EN_CTRL);
 
 	XPCS_SET_VAL(val, PMA_AFE_DFE_EN_CTRL, AFE_EN_0,
-		     pdata->mode_cfg->afe_en);
+		     serdes_cfg->afe_en[pdata->mode]);
 	XPCS_SET_VAL(val, PMA_AFE_DFE_EN_CTRL, AFE_EN_3_1,
 		     pdata->mode_cfg->afe_en_31);
 	XPCS_SET_VAL(val, PMA_AFE_DFE_EN_CTRL, DFE_EN_0,
-		     pdata->mode_cfg->dfe_en);
+		     serdes_cfg->dfe_en[pdata->mode]);
 	XPCS_SET_VAL(val, PMA_AFE_DFE_EN_CTRL, DFE_EN_3_1,
 		     pdata->mode_cfg->dfe_en_31);
 
 	XPCS_RGWR(pdata, PMA_AFE_DFE_EN_CTRL, val);
 
 	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL4, CONT_ADAPT_0,
-		      pdata->mode_cfg->cont_adapt0);
+		      serdes_cfg->cont_adapt0[pdata->mode]);
 	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL4, CONT_ADAPT_3_1,
 		      pdata->mode_cfg->cont_adapt31);
 
@@ -537,16 +606,16 @@ static void xpcs_cfg_table(struct xpcs_prv_data *pdata)
 	tx_eq0 = XPCS_RGRD(pdata, PMA_TX_EQ_C0);
 
 	XPCS_SET_VAL(tx_eq0, PMA_TX_EQ_C0, TX_EQ_MAIN,
-		     pdata->mode_cfg->tx_eq_main);
+		     serdes_cfg->tx_eq_main[pdata->mode]);
 	XPCS_SET_VAL(tx_eq0, PMA_TX_EQ_C0, TX_EQ_PRE,
-		     pdata->mode_cfg->tx_eq_pre);
+		     serdes_cfg->tx_eq_pre[pdata->mode]);
 
 	XPCS_RGWR(pdata, PMA_TX_EQ_C0, tx_eq0);
 
 	tx_eq1 = XPCS_RGRD(pdata, PMA_TX_EQ_C1);
 
 	XPCS_SET_VAL(tx_eq1, PMA_TX_EQ_C1, TX_EQ_POST,
-		     pdata->mode_cfg->tx_eq_post);
+		     serdes_cfg->tx_eq_post[pdata->mode]);
 	XPCS_SET_VAL(tx_eq1, PMA_TX_EQ_C1, TX_EQ_OVR_RIDE,
 		     pdata->mode_cfg->tx_eq_ovrride);
 
@@ -558,12 +627,14 @@ static void xpcs_cfg_table(struct xpcs_prv_data *pdata)
 	XPCS_RGWR_VAL(pdata, PMA_TX_GENCTRL1, VBOOST_EN_0, 1);
 
 	XPCS_RGWR_VAL(pdata, PMA_TX_GENCTRL1, VBOOST_LVL,
-		      pdata->mode_cfg->vboost_lvl);
+		      serdes_cfg->tx_vboost_lvl[pdata->mode]);
 }
 
 /* Switch to 10G XAUI Mode in 12G gen5 PHY */
 static int xpcs_teng_xaui_mode(struct xpcs_prv_data *pdata)
 {
+	struct xpcs_serdes_cfg *serdes_cfg = pdata->serdes_cfg;
+
 	if (xpcs_byp_pwrupseq(pdata) != XPCS_SUCCESS)
 		return XPCS_FAILURE;
 
@@ -599,10 +670,18 @@ static int xpcs_teng_xaui_mode(struct xpcs_prv_data *pdata)
 	XPCS_RGWR(pdata, PMA_RX_GENCTRL3, pdata->mode_cfg->los_thr);
 
 	/* PHY RX attenuation */
-	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL, pdata->serdes_cfg->rx_attn_lvl);
+	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL,
+		  serdes_cfg->rx_attn_lvl[pdata->mode]);
 
 	/* PHY ctle_pole and boost and gain register */
-	XPCS_RGWR(pdata, PMA_RX_EQ_CTRL0, pdata->mode_cfg->phy_boost_gain_val);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_BOOST_0,
+		      serdes_cfg->rx_ctle_boost[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_POLE_0,
+		      serdes_cfg->rx_ctle_pole[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA2_GAIN_0,
+		      serdes_cfg->rx_vga2_gain[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA1_GAIN_0,
+		      serdes_cfg->rx_vga1_gain[pdata->mode]);
 
 	/* Rx Biasing Current Control for Rx analog front end */
 	XPCS_RGWR_VAL(pdata, PMA_MISC_C0, RX_VREF_CTRL,
@@ -625,6 +704,8 @@ static int xpcs_teng_xaui_mode(struct xpcs_prv_data *pdata)
 /* Switch to 10G KR Mode in 12G gen5 PHY */
 static int xpcs_teng_kr_mode(struct xpcs_prv_data *pdata)
 {
+	struct xpcs_serdes_cfg *serdes_cfg = pdata->serdes_cfg;
+
 	if (xpcs_byp_pwrupseq(pdata) != XPCS_SUCCESS)
 		return XPCS_FAILURE;
 
@@ -650,10 +731,18 @@ static int xpcs_teng_kr_mode(struct xpcs_prv_data *pdata)
 	XPCS_RGWR(pdata, PMA_RX_GENCTRL3, pdata->mode_cfg->los_thr);
 
 	/* PHY RX attenuation level register */
-	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL, pdata->serdes_cfg->rx_attn_lvl);
+	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL,
+		  serdes_cfg->rx_attn_lvl[pdata->mode]);
 
 	/* PHY ctle_pole and boost and gain register */
-	XPCS_RGWR(pdata, PMA_RX_EQ_CTRL0, pdata->mode_cfg->phy_boost_gain_val);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_BOOST_0,
+		      serdes_cfg->rx_ctle_boost[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_POLE_0,
+		      serdes_cfg->rx_ctle_pole[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA2_GAIN_0,
+		      serdes_cfg->rx_vga2_gain[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA1_GAIN_0,
+		      serdes_cfg->rx_vga1_gain[pdata->mode]);
 
 	XPCS_RGWR(pdata, PMA_REF_CLK_CTRL, pdata->mode_cfg->ref_clk_ctrl);
 
@@ -683,6 +772,8 @@ static int xpcs_teng_kr_mode(struct xpcs_prv_data *pdata)
 /* 1G XAUI Mode */
 static int xpcs_oneg_xaui_mode(struct xpcs_prv_data *pdata)
 {
+	struct xpcs_serdes_cfg *serdes_cfg = pdata->serdes_cfg;
+
 	pdata->sgmii_type = PHY_SIDE_SGMII;
 
 	if (xpcs_byp_pwrupseq(pdata) != XPCS_SUCCESS)
@@ -721,10 +812,18 @@ static int xpcs_oneg_xaui_mode(struct xpcs_prv_data *pdata)
 	XPCS_RGWR(pdata, PMA_RX_GENCTRL3, pdata->mode_cfg->los_thr);
 
 	/* PHY RX attenuation level register */
-	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL, pdata->serdes_cfg->rx_attn_lvl);
+	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL,
+		  serdes_cfg->rx_attn_lvl[pdata->mode]);
 
 	/* PHY ctle_pole and boost and gain register */
-	XPCS_RGWR(pdata, PMA_RX_EQ_CTRL0, pdata->mode_cfg->phy_boost_gain_val);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_BOOST_0,
+		      serdes_cfg->rx_ctle_boost[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_POLE_0,
+		      serdes_cfg->rx_ctle_pole[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA2_GAIN_0,
+		      serdes_cfg->rx_vga2_gain[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA1_GAIN_0,
+		      serdes_cfg->rx_vga1_gain[pdata->mode]);
 
 	/* Rx Biasing Current Control for Rx analog front end */
 	XPCS_RGWR_VAL(pdata, PMA_MISC_C0, RX_VREF_CTRL,
@@ -750,6 +849,8 @@ static int xpcs_oneg_xaui_mode(struct xpcs_prv_data *pdata)
 /* 2.5G SGMII Mode */
 static int xpcs_2p5g_sgmii_mode(struct xpcs_prv_data *pdata)
 {
+	struct xpcs_serdes_cfg *serdes_cfg = pdata->serdes_cfg;
+
 	if (!strcmp(pdata->name, "wan_xpcs"))
 		pdata->mpllb = 0;
 	else
@@ -803,8 +904,19 @@ static int xpcs_2p5g_sgmii_mode(struct xpcs_prv_data *pdata)
 	/* PHY LOS threshold register */
 	XPCS_RGWR(pdata, PMA_RX_GENCTRL3, pdata->mode_cfg->los_thr);
 
+	/* PHY RX attenuation level register */
+	XPCS_RGWR(pdata, PMA_RX_ATTN_CTRL,
+		  serdes_cfg->rx_attn_lvl[pdata->mode]);
+
 	/* PHY ctle_pole and boost and gain register */
-	XPCS_RGWR(pdata, PMA_RX_EQ_CTRL0, pdata->mode_cfg->phy_boost_gain_val);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_BOOST_0,
+		      serdes_cfg->rx_ctle_boost[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, CTLE_POLE_0,
+		      serdes_cfg->rx_ctle_pole[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA2_GAIN_0,
+		      serdes_cfg->rx_vga2_gain[pdata->mode]);
+	XPCS_RGWR_VAL(pdata, PMA_RX_EQ_CTRL0, VGA1_GAIN_0,
+		      serdes_cfg->rx_vga1_gain[pdata->mode]);
 
 	/* Rx Biasing Current Control for Rx analog front end */
 	XPCS_RGWR_VAL(pdata, PMA_MISC_C0, RX_VREF_CTRL,
@@ -923,6 +1035,221 @@ static void xpcs_cl72_startup(struct xpcs_prv_data *pdata)
 	XPCS_RGWR_VAL(pdata, PMA_KR_PMD_CTRL, TR_EN, 1);
 }
 
+static int xpcs_get_serdes_cfg(struct xpcs_prv_data *pdata)
+{
+	struct xpcs_serdes_cfg *serdes_cfg;
+	int ret;
+
+	serdes_cfg = devm_kzalloc(pdata->dev, sizeof(*serdes_cfg), GFP_KERNEL);
+	if (!serdes_cfg)
+		return -ENOMEM;
+
+	if (device_property_present(pdata->dev, XPCS_AFE_EN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_AFE_EN,
+							 serdes_cfg->afe_en,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_AFE_EN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->afe_en, xpcs_serdes_afe_en,
+		       sizeof(serdes_cfg->afe_en));
+	}
+
+	if (device_property_present(pdata->dev, XPCS_DFE_EN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_DFE_EN,
+							 serdes_cfg->dfe_en,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_DFE_EN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->dfe_en, xpcs_serdes_dfe_en,
+		       sizeof(serdes_cfg->dfe_en));
+	}
+	if (device_property_present(pdata->dev, XPCS_ADAPT_EN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_ADAPT_EN,
+							 serdes_cfg->cont_adapt0,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_ADAPT_EN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->cont_adapt0, xpcs_serdes_cont_adapt_0,
+		       sizeof(serdes_cfg->cont_adapt0));
+	}
+	if (device_property_present(pdata->dev, XPCS_RX_ATTN_LVL)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_RX_ATTN_LVL,
+							 serdes_cfg->rx_attn_lvl,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_RX_ATTN_LVL);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->rx_attn_lvl, xpcs_serdes_rx_attn_ctrl,
+		       sizeof(serdes_cfg->rx_attn_lvl));
+	}
+	if (device_property_present(pdata->dev, XPCS_CTLE_BOOTS)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_CTLE_BOOTS,
+							 serdes_cfg->rx_ctle_boost,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_CTLE_BOOTS);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->rx_ctle_boost, xpcs_serdes_ctrl_boost_0,
+		       sizeof(serdes_cfg->rx_ctle_boost));
+	}
+	if (device_property_present(pdata->dev, XPCS_CTLE_POLE)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_CTLE_POLE,
+							 serdes_cfg->rx_ctle_pole,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_CTLE_POLE);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->rx_ctle_pole, xpcs_serdes_ctrl_pole_0,
+		       sizeof(serdes_cfg->rx_ctle_pole));
+	}
+	if (device_property_present(pdata->dev, XPCS_VGA2_GAIN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_VGA2_GAIN,
+							 serdes_cfg->rx_vga2_gain,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_VGA2_GAIN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->rx_vga2_gain, xpcs_serdes_vga2_gain_0,
+		       sizeof(serdes_cfg->rx_vga2_gain));
+	}
+	if (device_property_present(pdata->dev, XPCS_VGA1_GAIN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_VGA1_GAIN,
+							 serdes_cfg->rx_vga1_gain,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_VGA1_GAIN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->rx_vga1_gain, xpcs_serdes_vga1_gain_0,
+		       sizeof(serdes_cfg->rx_vga1_gain));
+	}
+	if (device_property_present(pdata->dev, XPCS_TX_EQ_PRE)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_TX_EQ_PRE,
+							 serdes_cfg->tx_eq_pre,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_TX_EQ_PRE);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->tx_eq_pre, xpcs_serdes_tx_eq_pre,
+		       sizeof(serdes_cfg->tx_eq_pre));
+	}
+	if (device_property_present(pdata->dev, XPCS_TX_EQ_POST)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_TX_EQ_POST,
+							 serdes_cfg->tx_eq_post,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_TX_EQ_POST);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->tx_eq_post, xpcs_serdes_tx_eq_post,
+		       sizeof(serdes_cfg->tx_eq_post));
+	}
+	if (device_property_present(pdata->dev, XPCS_TX_EQ_MAIN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_TX_EQ_MAIN,
+							 serdes_cfg->tx_eq_main,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_TX_EQ_MAIN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->tx_eq_main, xpcs_serdes_tx_eq_main,
+		       sizeof(serdes_cfg->tx_eq_main));
+	}
+	if (device_property_present(pdata->dev, XPCS_TX_IBOOST_LVL)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_TX_IBOOST_LVL,
+							 serdes_cfg->tx_iboost_lvl,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_TX_IBOOST_LVL);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->tx_iboost_lvl, xpcs_serdes_tx_iboost_lvl,
+		       sizeof(serdes_cfg->tx_iboost_lvl));
+	}
+	if (device_property_present(pdata->dev, XPCS_TX_VBOOST_LVL)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_TX_VBOOST_LVL,
+							 serdes_cfg->tx_vboost_lvl,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_TX_VBOOST_LVL);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->tx_vboost_lvl, xpcs_serdes_tx_vboost_lvl,
+		       sizeof(serdes_cfg->tx_vboost_lvl));
+	}
+	if (device_property_present(pdata->dev, XPCS_TX_VBOOST_EN)) {
+		ret = device_property_read_u32_array(pdata->dev,
+						     XPCS_TX_VBOOST_EN,
+							 serdes_cfg->tx_vboost_en,
+							 MAX_XPCS_MODE);
+		if (ret) {
+			dev_err(pdata->dev, "invalid %s property\n",
+				XPCS_TX_VBOOST_EN);
+			return ret;
+		}
+	} else {
+		memcpy(serdes_cfg->tx_vboost_en, xpcs_serdes_vboost_en,
+		       sizeof(serdes_cfg->tx_vboost_en));
+	}
+	serdes_cfg->no_phy_rx_auto_adapt = of_property_read_bool
+				(pdata->dev->of_node,
+				XPCS_NO_PHY_RX_AUTO_ADAPT);
+
+	pdata->serdes_cfg = serdes_cfg;
+
+	return 0;
+}
+
 static struct xpcs_mode_cfg *xpcs_get_cfg(u32 mode)
 {
 	int i = 0;
@@ -938,7 +1265,6 @@ static struct xpcs_mode_cfg *xpcs_get_cfg(u32 mode)
 static int xpcs_init(struct xpcs_prv_data *pdata)
 {
 	int ret = XPCS_SUCCESS;
-	struct xpcs_prv_data data_t = {0};
 
 	pdata->pcs_mode = BACKPL_ETH_PCS;
 
@@ -948,17 +1274,7 @@ static int xpcs_init(struct xpcs_prv_data *pdata)
 		return -EINVAL;
 	}
 
-	memcpy(&data_t, pdata, sizeof(struct xpcs_prv_data));
-
-	if (data_t.serdes_cfg->f_serdes & RX_ENQ) {
-		pdata->mode_cfg->afe_en = pdata->serdes_cfg->afe_en;
-		pdata->mode_cfg->dfe_en = pdata->serdes_cfg->dfe_en;
-		pdata->mode_cfg->cont_adapt0 = pdata->serdes_cfg->cont_adapt0;
-		pdata->mode_cfg->phy_boost_gain_val =
-					pdata->serdes_cfg->phy_boost_gain_val;
-	}
-
-	ret = pdata->mode_cfg->set_mode(&data_t);
+	ret = pdata->mode_cfg->set_mode(pdata);
 
 	if (ret != XPCS_SUCCESS)
 		dev_info(pdata->dev, "%s: %d :: Failed ::\n",
@@ -967,32 +1283,11 @@ static int xpcs_init(struct xpcs_prv_data *pdata)
 	return ret;
 }
 
-static int xpcs_update_cfg(struct device *dev, struct xpcs_mode_cfg *xpcs_cfg,
-			   char *mode_name)
-{
-	struct of_phandle_args args;
-
-	if (of_parse_phandle_with_fixed_args(dev->of_node, mode_name,
-					     3, 0, &args) < 0) {
-		pr_err("Cannot get config args from device tree\n");
-		return -1;
-	}
-
-	xpcs_cfg->tx_eq_pre = args.args[0];
-	xpcs_cfg->tx_eq_post = args.args[1];
-	xpcs_cfg->tx_eq_main = args.args[2];
-
-	return 0;
-}
-
 #ifdef CONFIG_OF
 static int xpcs_parse_dts(struct platform_device *pdev,
 			  struct xpcs_prv_data **pdata)
 {
-	struct device_node *serdes = pdev->dev.of_node;
-	struct device_node *rx;
 	struct device *dev = &pdev->dev;
-	struct xpcs_mode_cfg *xpcs_cfg = NULL;
 	struct of_phandle_args args;
 	int i;
 	u32 prop = 0;
@@ -1002,13 +1297,6 @@ static int xpcs_parse_dts(struct platform_device *pdev,
 	if (!(*pdata))
 		return -ENOMEM;
 
-	(*pdata)->serdes_cfg = devm_kzalloc(dev,
-					    sizeof(struct xpcs_serdes_cfg),
-					    GFP_KERNEL);
-
-	if (!(*pdata)->serdes_cfg)
-		return -ENOMEM;
-
 	/* Retrieve the xpcs name */
 	if (device_property_read_string(dev, XPCS_MOD_NAME, &(*pdata)->name)) {
 		dev_err(dev, "Xpcs name: cannot get property\n");
@@ -1050,48 +1338,6 @@ static int xpcs_parse_dts(struct platform_device *pdev,
 		return -EINVAL;
 	}
 
-	/* Retrieve XPCS SERDES configs */
-	serdes = of_get_child_by_name(serdes, XPCS_SERDES_CFG);
-	if (serdes) {
-		(*pdata)->serdes_cfg->no_phy_rx_auto_adapt =
-		       of_property_read_bool(serdes, XPCS_NO_PHY_RX_AUTO_ADAPT);
-		rx = of_get_child_by_name(serdes, "rx_enq");
-		if (rx) {
-			(*pdata)->serdes_cfg->f_serdes = RX_ENQ;
-			if (!of_property_read_u32(rx, XPCS_AFE_EN, &prop))
-				(*pdata)->serdes_cfg->afe_en = prop;
-			if (!of_property_read_u32(rx, XPCS_DFE_EN, &prop))
-				(*pdata)->serdes_cfg->dfe_en = prop;
-			if (!of_property_read_u32(rx, XPCS_ADAPT_EN, &prop))
-				(*pdata)->serdes_cfg->cont_adapt0 = prop;
-			if (!of_property_read_u32(rx, XPCS_RX_ATTN_LVL, &prop))
-				(*pdata)->serdes_cfg->rx_attn_lvl = prop;
-			if (!of_property_read_u32(rx, XPCS_CTLE_BOOTS, &prop))
-				(*pdata)->serdes_cfg->rx_ctle_boost = prop;
-			if (!of_property_read_u32(rx, XPCS_CTLE_POLE, &prop))
-				(*pdata)->serdes_cfg->rx_ctle_pole = prop;
-			if (!of_property_read_u32(rx, XPCS_VGA2_GAIN, &prop))
-				(*pdata)->serdes_cfg->rx_vga2_gain = prop;
-			if (!of_property_read_u32(rx, XPCS_VGA1_GAIN, &prop))
-				(*pdata)->serdes_cfg->rx_vga1_gain = prop;
-
-			prop = PHY_BOOST((*pdata)->serdes_cfg->rx_ctle_boost,
-					 (*pdata)->serdes_cfg->rx_ctle_pole,
-					 (*pdata)->serdes_cfg->rx_vga2_gain,
-					 (*pdata)->serdes_cfg->rx_vga1_gain);
-			(*pdata)->serdes_cfg->phy_boost_gain_val = prop;
-		}
-	}
-
-	xpcs_cfg = xpcs_get_cfg(ONEG_XAUI_MODE);
-	xpcs_update_cfg(dev, xpcs_cfg, "serdes_1g_cfg");
-
-	xpcs_cfg = xpcs_get_cfg(TWOP5G_SGMII_MODE);
-	xpcs_update_cfg(dev, xpcs_cfg, "serdes_1g_cfg");
-
-	xpcs_cfg = xpcs_get_cfg(TENG_KR_MODE);
-	xpcs_update_cfg(dev, xpcs_cfg, "serdes_10g_cfg");
-
 	/* Retrieve the dynamic link cfg sequence */
 	if (of_parse_phandle_with_fixed_args(dev->of_node,
 	    LINK_MODE_CFG, MAX_XPCS, 0, &args) < 0)
@@ -1170,6 +1416,7 @@ static int xpcs_power_off(struct device *dev)
 static int xpcs_power_on(struct device *dev)
 {
 	struct xpcs_prv_data *pdata;
+	struct xpcs_serdes_cfg *serdes_cfg;
 	struct phy *phy;
 	int ret = 0;
 
@@ -1177,6 +1424,7 @@ static int xpcs_power_on(struct device *dev)
 		return -ENODEV;
 	dev_dbg(dev, "%s\n", __func__);
 	pdata = dev_get_drvdata(dev);
+	serdes_cfg = pdata->serdes_cfg;
 	phy = pdata->phy;
 	if (!phy) {
 		dev_err(dev, "XPCS Power On: Cannot get phy!\n");
@@ -1209,7 +1457,7 @@ static int xpcs_power_on(struct device *dev)
 
 	pdata->state = 1;
 
-	if (!pdata->serdes_cfg->no_phy_rx_auto_adapt)
+	if (!serdes_cfg->no_phy_rx_auto_adapt)
 		ret = phy_calibrate(phy);
 
 	return ret;
@@ -1358,6 +1606,7 @@ static int xpcs_probe(struct platform_device *pdev)
 	struct device *dev = &pdev->dev;
 	int ret = XPCS_SUCCESS;
 	struct xpcs_prv_data *pdata;
+	struct xpcs_serdes_cfg *serdes_cfg;
 
 	if (dev->of_node) {
 		if (xpcs_parse_dts(pdev, &pdata) != XPCS_SUCCESS) {
@@ -1431,6 +1680,12 @@ static int xpcs_probe(struct platform_device *pdev)
 		return -EINVAL;
 	}
 
+	ret = xpcs_get_serdes_cfg(pdata);
+	if (ret)
+		return ret;
+
+	serdes_cfg = pdata->serdes_cfg;
+
 	/* Initialize XPCS */
 	if (xpcs_init(pdata)) {
 		dev_err(dev, "%s Initialization Failed!!\n", pdata->name);
@@ -1445,7 +1700,7 @@ static int xpcs_probe(struct platform_device *pdev)
 	}
 
 	/* Phy calibrate for RX Signal Adaption */
-	if (!pdata->serdes_cfg->no_phy_rx_auto_adapt)
+	if (!serdes_cfg->no_phy_rx_auto_adapt)
 		ret = phy_calibrate(pdata->phy);
 
 	/* Schedule the link poll task if link is dynmically configurable*/
diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.h b/drivers/net/ethernet/lantiq/xpcs/xpcs.h
index 96ef7c742f04..6f99cc000955 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.h
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.h
@@ -553,6 +553,15 @@
 #define PCS_DIG_STS_PSEQ_STATE_POS		2
 #define PCS_DIG_STS_PSEQ_STATE_WIDTH		3
 
+#define PMA_RX_EQ_CTRL0_CTLE_BOOST_0_POS	0
+#define PMA_RX_EQ_CTRL0_CTLE_BOOST_0_WIDTH	4
+#define PMA_RX_EQ_CTRL0_CTLE_POLE_0_POS		5
+#define PMA_RX_EQ_CTRL0_CTLE_POLE_0_WIDTH	3
+#define PMA_RX_EQ_CTRL0_VGA2_GAIN_0_POS	8
+#define PMA_RX_EQ_CTRL0_VGA2_GAIN_0_WIDTH	4
+#define PMA_RX_EQ_CTRL0_VGA1_GAIN_0_POS	12
+#define PMA_RX_EQ_CTRL0_VGA1_GAIN_0_WIDTH	4
+
 #define MAX_XPCS				6
 
 enum {
@@ -613,11 +622,6 @@ enum {
 	PCS_TYPE_RESERVED,
 };
 
-enum {
-	RX_ENQ = 1,
-	TX_ENQ
-};
-
 enum {
 	 SGMII_MODE_1000BASET = 4,
 	 SGMII_MODE_2500BASET,
@@ -638,11 +642,8 @@ struct xpcs_mode_cfg {
 	u32 mpllb_bw;
 	u32 vco_ld_val[4];
 	u32 vco_ref_ld[4];
-	u32 afe_en;
 	u32 afe_en_31;
-	u32 dfe_en;
 	u32 dfe_en_31;
-	u32 cont_adapt0;
 	u32 cont_adapt31;
 	u32 tx_rate[4];
 	u32 rx_rate[4];
@@ -656,32 +657,110 @@ struct xpcs_mode_cfg {
 	u32 mpllb_div10_clk_en;
 	u32 mpllb_div8_clk_en;
 	u32 rx_div165_clk_en;
-	u32 tx_eq_main;
-	u32 tx_eq_pre;
-	u32 tx_eq_post;
 	u32 tx_eq_ovrride;
 	u32 los_thr;
-	u32 phy_boost_gain_val;
 	u32 rx_vref_ctrl;
 	u32 ref_clk_ctrl;
-	u32 vboost_lvl;
 	u32 tx_iboost[4];
 	int (*set_mode)(struct xpcs_prv_data *);
 };
 
-/* Port specific XPCS SERDES cfg */
+/* RX EQ XPCS Serdes Default ettings */
+#define XPCS_TENG_KR_AFE_EN 0x1
+#define XPCS_TENG_XAUI_AFE_EN 0x1
+#define XPCS_ONEG_AFE_EN 0x0
+#define XPCS_TWOP5G_AFE_EN 0x0
+
+#define XPCS_TENG_KR_DFE_EN 0x1
+#define XPCS_TENG_XAUI_DFE_EN 0x1
+#define XPCS_ONEG_DFE_EN 0x0
+#define XPCS_TWOP5G_DFE_EN 0x0
+
+#define XPCS_TENG_KR_CONT_ADAPT_0 0x1
+#define XPCS_TENG_XAUI_CONT_ADAPT_0 0x1
+#define XPCS_ONEG_CONT_ADAPT_0 0x0
+#define XPCS_TWOP5G_CONT_ADAPT_0 0x0
+
+#define XPCS_TENG_KR_PMA_RX_ATTN_CTRL 0x7
+#define XPCS_TENG_XAUI_PMA_RX_ATTN_CTRL 0x7
+#define XPCS_ONEG_PMA_RX_ATTN_CTRL 0x7
+#define XPCS_TWOP5G_PMA_RX_ATTN_CTRL 0x7
+
+#define XPCS_TENG_KR_PMA_RX_CTLE_BOOST_0 0xA
+#define XPCS_TENG_XAUI_PMA_RX_CTLE_BOOST_0 0xA
+#define XPCS_ONEG_PMA_RX_CTLE_BOOST_0 0x6
+#define XPCS_TWOP5G_PMA_RX_CTLE_BOOST_0 0x6
+
+#define XPCS_TENG_KR_PMA_RX_EQ_CTRL0 0x4
+#define XPCS_TENG_XAUI_PMA_RX_EQ_CTRL0 0x4
+#define XPCS_TWOP5G_PMA_RX_EQ_CTRL0 0x0
+#define XPCS_ONEG_PMA_RX_EQ_CTRL0 0x0
+
+#define XPCS_TENG_KR_PMA_RX_VGA2_GAIN_0 0x7
+#define XPCS_TENG_XAUI_PMA_RX_VGA2_GAIN_0 0x7
+#define XPCS_TWOP5G_PMA_RX_VGA2_GAIN_0 0x7
+#define XPCS_ONEG_PMA_RX_VGA2_GAIN_0 0x7
+
+#define XPCS_TENG_KR_PMA_RX_VGA1_GAIN_0 0x7
+#define XPCS_TENG_XAUI_PMA_RX_VGA1_GAIN_0 0x7
+#define XPCS_TWOP5G_PMA_RX_VGA1_GAIN_0 0x7
+#define XPCS_ONEG_PMA_RX_VGA1_GAIN_0 0x7
+
+#define XPCS_TENG_KR_PMA_TX_EQ_MAIN 0x24
+#define XPCS_TENG_XAUI_PMA_TX_EQ_MAIN 0x24
+#define XPCS_TWOP5G_PMA_TX_EQ_MAIN 0X13
+#define XPCS_ONEG_PMA_TX_EQ_MAIN 0x10
+
+#define XPCS_TENG_KR_PMA_TX_EQ_PRE 0x10
+#define XPCS_TENG_XAUI_PMA_TX_EQ_PRE 0x10
+#define XPCS_TWOP5G_PMA_TX_EQ_PRE 0X4
+#define XPCS_ONEG_PMA_TX_EQ_PRE 0x4
+
+#define XPCS_TENG_KR_PMA_TX_EQ_POST 0x0
+#define XPCS_TENG_XAUI_PMA_TX_EQ_POST 0x0
+#define XPCS_TWOP5G_PMA_TX_EQ_POST 0X4
+#define XPCS_ONEG_PMA_TX_EQ_POST 0x4
+
+#define XPCS_TENG_KR_PMA_TX_EQ_IBOOST_LVL 0xF
+#define XPCS_TENG_XAUI_PMA_TX_EQ_IBOOST_LVL 0xF
+#define XPCS_TWOP5G_PMA_TX_EQ_IBOOST_LVL 0xF
+#define XPCS_ONEG_PMA_TX_EQ_IBOOST_LVL 0xF
+
+#define XPCS_TENG_KR_PMA_TX_EQ_VBOOST_LVL 0x5
+#define XPCS_TENG_XAUI_PMA_TX_EQ_VBOOST_LVL 0x5
+#define XPCS_TWOP5G_PMA_TX_EQ_VBOOST_LVL 0x5
+#define XPCS_ONEG_PMA_TX_EQ_VBOOST_LVL 0x5
+
+#define XPCS_TENG_KR_PMA_TX_EQ_VBOOST_EN 0x1
+#define XPCS_TENG_XAUI_PMA_TX_EQ_VBOOST_EN 0x1
+#define XPCS_TWOP5G_PMA_TX_EQ_VBOOST_EN 0x1
+#define XPCS_ONEG_PMA_TX_EQ_VBOOST_EN 0x1
+
 struct xpcs_serdes_cfg {
-	u32 f_serdes;
-	u32 no_phy_rx_auto_adapt;
-	u32 afe_en;
-	u32 dfe_en;
-	u32 cont_adapt0;
-	u32 rx_attn_lvl;
-	u32 rx_ctle_boost;
-	u32 rx_ctle_pole;
-	u32 rx_vga2_gain;
-	u32 rx_vga1_gain;
-	u32 phy_boost_gain_val;
+	/* TX/RX EQ XPCS Serdes configurable settings
+	 * related to the transmission line loss
+	 * on the PCB board design,
+	 * Switching beetween speeds require new serdes
+	 * configurable settings, the required values can be
+	 * supplied as device properties in array format
+	 * Entry sequence is 10GbE KR, 10GbE XAUI, 1GbE and 2.5GbE.
+	 */
+	u32 afe_en[MAX_XPCS_MODE];
+	u32 dfe_en[MAX_XPCS_MODE];
+	u32 cont_adapt0[MAX_XPCS_MODE];
+	u32 rx_attn_lvl[MAX_XPCS_MODE];
+	u32 rx_ctle_boost[MAX_XPCS_MODE];
+	u32 rx_ctle_pole[MAX_XPCS_MODE];
+	u32 rx_vga2_gain[MAX_XPCS_MODE];
+	u32 rx_vga1_gain[MAX_XPCS_MODE];
+	u32 tx_eq_main[MAX_XPCS_MODE];
+	u32 tx_eq_pre[MAX_XPCS_MODE];
+	u32 tx_eq_post[MAX_XPCS_MODE];
+	u32 tx_iboost_lvl[MAX_XPCS_MODE];
+	u32 tx_vboost_lvl[MAX_XPCS_MODE];
+	u32 tx_vboost_en[MAX_XPCS_MODE];
+	/* Phy Rx auto adapt */
+	bool no_phy_rx_auto_adapt;
 };
 
 struct xpcs_ops {
@@ -738,9 +817,6 @@ struct xpcs_prv_data {
 	/* XPCS Mode Switching */
 	struct xpcs_mode_cfg *mode_cfg;
 
-	/* XPCS SERDES cfg */
-	struct xpcs_serdes_cfg *serdes_cfg;
-
 	/* Scheduled link poll workitem */
 	struct workqueue_struct *link_poll;
 	/* Workqueue name */
@@ -755,6 +831,9 @@ struct xpcs_prv_data {
 	unsigned int poll_interval;
 	/* Link is Dynamic or static */
 	bool link_cfg;
+
+	/* Serdes configurable data */
+	void *serdes_cfg;
 };
 
 #define xpcs_r32(reg)		readl(reg)
